### PR TITLE
Add arguments to transpile

### PR DESCRIPTION
**Pull request by hnformentin**
Solves #308

### CODE REVIEW

**Comment by kwinkunks** [comment]
I had a thought about how to add the urgently needed capability to produce a single 'cell' in a release matrix, without having to hardcode 'redhat' and 'python' into the tool in yet another place.

Maybe the interface could look like:

    komodo-transpiler transpile
        --matrix-file=releases/matrices/${{ inputs.target }}.yml
        --job py38,rhel7
        --output releases

This `--job` argument can then be split into the two 'coordinates' of the matrix. We are free to interpret them however, but I think the tool only needs to care that they are 'thing 1' and 'thing 2', right?

Maybe `release_transpiler.py` can also be refactored to not use variable names like `rhel_ver` and `py_ver`, but I didn't look at how hard that would be. And eventually we will want it to allow for an arbitrary number of 'coordinates' in `job` but we can deal with that when we need it.

What do you think @xjules and @hnformentin? Would this work?

**Comment by kwinkunks** [comment]
:point_up: Maybe I should move this comment to the issue??

File: komodo/matrix.py
Comment (kwinkunks) [inline komodo/matrix.py:26]: I wonder if we really need to hardcode details of our particular matrix into here? For example, can we define a more general pattern that ~all release names will adhere to?

Maybe releases will all look like `1111.11.11a11-pppp11-os11-other111`? At least that would avoid mentioning `py` and `rhel`. (Or maybe there's some other pattern, this is just an example.) The beauty of regex is that we can avoid having to be too specific. At its simplest, maybe it's just "everything before the first hyphen".

Alternatively, maybe the details of the matrix can go in a site config file? Or get passed in to the CLI that's making this query?

File: komodo/matrix.py
Comment (kwinkunks) [inline komodo/matrix.py:28]: You should be able to use capturing groups (using parentheses) and more specific wildcards, like `[0-9]` or `\d` for digits, to capture everything 'at once' here. Something like `\d{2,3}` means "2 or 3 digits", for example. 

Here's one of those sites you can practise regex on: https://regex101.com/ -- and we can take a look together if you want.

File: komodo/matrix.py
Comment (hnformentin) [reply komodo/matrix.py:26]: Thanks for the comment. 
The format of a release is governed by [format_release](https://github.com/equinor/komodo/blob/f20735c60b84620ee3d08b15afd8f05331e30ddb/komodo/matrix.py#L19) function. I changed the code to keep the link with that function, which is a little better than pure hardcoded.
Nevertheless, I cannot see how we could get rid of hardcoding completely.

There is [one test](https://github.com/equinor/komodo/blob/f8fc14e46b0cfbc958d0ccb17a190dcbadd55fab/tests/test_matrix.py#L17) that would not pass with the `-` rule.

On the alternative you mention, do you suggest to have another argument to `transpile`? For example, we would have 4 arguments (or 2) for the current coordinates:  separate `os` from `os_version` and `prog_language` from `prog_language_version`?

File: komodo/matrix.py
Comment (hnformentin) [reply komodo/matrix.py:28]: Excellent tip!

**Comment by hnformentin** [comment]
I added some commits with improvements from the review version: change order, remove repetition, fix black. 
Will rebase all in one commit before merging.

File: komodo/release_transpiler.py
Comment (kwinkunks) [inline komodo/release_transpiler.py:15]: My thought was that by using the `coords` dict, I think it should be possible to get rid of `py_ver` and `rhel_ver` completely. Instread of passing those variable around, pass the `coords` dict around. Maybe I'm missing something and it's not possible, but it seems like the `coords` dict should be able to have entries like `macos_ver:['10.15', '10.16']` and the code will just accommodate it as another constraint to add to the release string. This would mean looping over both the keys and the values to generate all the various release name strings.

(I _think_ this would mean we could also do things like this with another entry in the `coords` dict > https://github.com/equinor/komodo/pull/331 )

**Review by kwinkunks — CHANGES_REQUESTED** [review]

File: komodo/release_transpiler.py
Comment (hnformentin) [reply komodo/release_transpiler.py:106]: If I understand you correctly, we actually had this default in `matrix.py` as 
```
RH_VERSIONS = ["7"]
PY_VERSIONS = ["3.8"]
```
and this PR is substituting it to a single place, as you suggested in the last part of the comment.
Did I misunderstand something?

File: komodo/release_transpiler.py
Comment (hnformentin) [reply komodo/release_transpiler.py:15]: I fully understand your point.
In order to keep the scope of the issue, I would prefer to use another issue i.e. #332 to consider turning komodo agnostic to os and dev system.

**Comment by kwinkunks** [comment]
I think it's worth adding the type of the return to the functions as well. You should also add docstrings -- now that you've taken the time to figure out what these things do, I think it will help the next person. (Is the type of the YAML dict really `OrderedDict`? I would have thought `dict` would do, since the basic type is ordered since Py 3.6. Maybe just check the type or see what MyPy says.)

I'm still bothered by the very specific regex in `matrix.py`, which looks for Python and RHEL. Is it really not possible to generalize the pattern? Or, perhaps better, pass in the matrix itself (`coords` I guess, see below) and figure out the 'base' from there? And again, you can use parentheses to capture the group you want and avoid needing `re.split()`. 

I've a feeling `coords` should be called `matrix`, what do you think?

**Comment by hnformentin** [comment]
> I think it's worth adding the type of the return to the functions as well. You should also add docstrings -- now that you've taken the time to figure out what these things do, I think it will help the next person. 

Nice, I tried to add docstring and types of returns. Hope it helps the next to come across the code.

> (Is the type of the YAML dict really `OrderedDict`? I would have thought `dict` would do, since the basic type is ordered since Py 3.6. Maybe just check the type or see what MyPy says.)

I checked the type within a test and it is indeed OrderedDict. Also tried mypy (which turned up to be useful to identify an non-imported package), and it seems to be fine with it.

> I'm still bothered by the very specific regex in `matrix.py`, which looks for Python and RHEL. Is it really not possible to generalize the pattern? Or, perhaps better, pass in the matrix itself (`coords` I guess, see below) and figure out the 'base' from there? And again, you can use parentheses to capture the group you want and avoid needing `re.split()`.

For sure there is way to remove and generalise the pattern, but as it is not the focus of this issue, I will let this work to the issue #332 . I refer the issue so the one picking it can access the comments/suggestions here.

> I've a feeling `coords` should be called `matrix`, what do you think?

I think `coords` is slightly better because calling `matrix` would potentially generate confusion if it is the matrix file, for example. So I would stick with `coords`. Maybe `axis` would be another possibility, or `row` and `column`, though this last option could be even more confusing!

**Comment by hnformentin** [comment]
Aiming to remove the dependence of `combine` of the hard-coded PY_VERSION variable that we had in `matrix.py`, I ended up choosing to specify the argument `--py_coord` -  a list of python versions to be considered. This is an optional arg, and if it is not defined, we will search which are the files in the release_folder and combine all of those that match the release_base.

File: komodo/release_transpiler.py
Comment (kwinkunks) [inline komodo/release_transpiler.py:183]: I thought the Python version (and any arbitrary matrix 'coordinates') could be encoded in the `coords` argument?

File: komodo/release_transpiler.py
Comment (kwinkunks) [inline komodo/release_transpiler.py:79]: Does this definitely need to be of type `OrderedDict`? It seems like `dict` should suffice. I think the object is coming from `load_yaml`, which I think uses `ruamel`, but maybe just check its type, and/or see if you can 'get away' with just a dict (which will be ordered anyway).

File: komodo/release_transpiler.py
Comment (kwinkunks) [inline komodo/release_transpiler.py:62]: The function also takes an `rhel` version, so maybe the docstring should reflect that.

Did you try just passing in arbitrary coordinates here? I still think it should be possible to make this function agnostic to the matrix. (E.g. have it consume `coords` instead of `rhel_ver` and `py_ver` and loop over its keys.)

File: komodo/release_transpiler.py
Comment (hnformentin) [reply komodo/release_transpiler.py:184]: As the code of the function `combine` is of now, it needs only the python versions.
I guess this is due to historical reasons.
The options were:
1. keep `PY_VERSION=[3.8]` hardcoded as it was previously;
2. pass `coords` even though `rhel_ver` is not used, which means we are passing an unnecessary argument;
3. let only `--py-coords` as an optional argument;
4. remove the function `combine` as a whole - it is not used by us now. I preferred not because it maybe an useful feature for users outside scout or maybe can be useful in the future.

I preferred the third option, but would you prefer the second one?

File: komodo/release_transpiler.py
Comment (hnformentin) [reply komodo/release_transpiler.py]: Yes, indeed `OrderedDict` could be substituted overall. Thanks!

**Comment by hnformentin** [comment]
I cannot see the specific changed you required, from `Merging can be performed automatically once the requested changes are addressed`. Can you please point out them into a comment or help me to find them otherwise?

File: tests/test_release_transpiler.py
Comment (kwinkunks) [inline tests/test_release_transpiler.py:40]: Do the tests produce the correct results if you use (say) `{"py": ["3.8", "3.10"], "rhel": ["7", "8"]}`?

File: komodo/release_transpiler.py
Comment (kwinkunks) [reply komodo/release_transpiler.py:184]: I think we should avoid adding Equinor-specific arguments so I think it should accept the same matrix definition (i.e. `coords` --- which should maybe be called `matrix`).

If we don't want to refactor it now because we don't use it, maybe we should deprecate it.

File: komodo/matrix.py
Comment (kwinkunks) [reply komodo/matrix.py:27]: Eventually I think `format_release()` could look more like this:

```python
def format_release(base, matrix):
    for coords in product(*matrix.values()):
        coord_strings = [f"{name}{coord}" for name, coord in zip(matrix.keys(), coords)]
        yield '-'.join([base] + coord_strings)
```

Then be called like...

```python
matrix = {"py": ["3.8", "3.10", "3.11"], "rhel": ["7"], "loc": ["Azure", "Onprem"]}

for fname in format_release('baseybase', matrix):
    print(fname)
```

This produces:

```
baseybase-py3.8-rhel7-locAzure
baseybase-py3.8-rhel7-locOnprem
baseybase-py3.10-rhel7-locAzure
baseybase-py3.10-rhel7-locOnprem
baseybase-py3.11-rhel7-locAzure
baseybase-py3.11-rhel7-locOnprem
```

This way, it feels like we have lots more flexibility because the code never needs to know what our matrix names and coordinates are --- we could add Ruby or switch to Ubuntu without changing anything.

If this is all too much to ask of this PR, no worries --- I just wanted to clarify what I had in my mind! (For me as much as anything :)

File: tests/test_release_transpiler.py
Comment (hnformentin) [reply tests/test_release_transpiler.py]: yeap, I included the suggestion in the test and updated the test data `test_release_matrix.yml` for that.

File: komodo/release_transpiler.py
Comment (kwinkunks) [reply komodo/release_transpiler.py]: Okay, if what you have does what's needed then :+1:

File: komodo/release_transpiler.py
Comment (kwinkunks) [reply komodo/release_transpiler.py:184]: Is it possible to avoid adding another specific 'coordinate' argument here?

File: komodo/release_transpiler.py
Comment (kwinkunks) [reply komodo/release_transpiler.py]: Right -- I think this point is out of date wrt the code, but either way, I will just focus on #332 for this sort of thing.

File: komodo/release_transpiler.py
Comment (kwinkunks) [reply komodo/release_transpiler.py]: I think this is outdated; will double-check in latest.

File: komodo/matrix.py
Comment (kwinkunks) [reply komodo/matrix.py:27]: If this won't change, then I'll let you hit *Resolve conversation* to close this thread.

**Comment by kwinkunks** [comment]
> I cannot see the specific changed you required, from `Merging can be performed automatically once the requested changes are addressed`. Can you please point out them into a comment or help me to find them otherwise?

Not totally sure but I _think_ it's referring to the **Resolve conversation** items in the review. I think it's okay for you to close those on things as they are resolved (or rejected!).

File: komodo/release_transpiler.py
Comment (hnformentin) [reply komodo/release_transpiler.py:184]: > Is it possible to avoid adding another specific 'coordinate' argument here?

I understand you refer to 'coordinate' to `--py_coords` argument. If so, it is an argument keep the previous functionaty of `combine`. But it is optional and by default the versions to be combined will be retrieved in the `release_folder`.

As it is an optional argument, I don't see the problem of the solution. Do you have something particularly in mind?
Remove the option and let only the default behaviour?

File: komodo/matrix.py
Comment (hnformentin) [reply komodo/matrix.py:27]: That is indeed a good point. I renamed the entry argument as `matrix` and referred your specific suggestion in #332 . Will certainly help who picks that issue.

File: komodo/release_transpiler.py
Comment (kwinkunks) [reply komodo/release_transpiler.py:184]: If there's really no way to avoid adding an argument to the interface that we will have to deprecate when we eventually fix all this Py/RHEL stuff, then it has to be done. I trust that you've considered all the alternatives.

**Comment by hnformentin** [comment]
For reference, the feature to make komodo more agnostic and let the argument `matrix` to be more flexible was discussed a couple of times within the squad. As it is referred in some comments in this PR, we decided to make another issue to enable this future proof flexibility #332 as this was not the priority at the moment. Thanks for the comments in the review so far @kwinkunks 🤓

File: komodo/release_transpiler.py
Comment (valentin-krasontovitsch) [inline komodo/release_transpiler.py:61]: can we rename this function? the idea is that we based on rh and py versions pick a specific version for each package, right? what do you think about:

```suggestion
def _package_version_picker(packages: dict, rhel_ver: str, py_ver: str) -> dict:
```

File: komodo/release_transpiler.py
Comment (valentin-krasontovitsch) [inline komodo/release_transpiler.py:79]: yet another renaming suggestion. the versions in the function name seem to point to the versions argument, but we're not checking that those are consistent - rather we're checking that the given "coordinates" (rhel version + py version) are present in the package versions... thankfully this is so simple XD
how about something like

```suggestion
def _check_version_exists_for_coordinates(
```

and perhaps replace `rhel_ver` with `rhel_coordinate`

but this is maybe just bike-shedding, i dunno, the naming in this context is difficult, as i feel like we don't really have conceptual names down for what we're working with...

File: komodo/release_transpiler.py
Comment (valentin-krasontovitsch) [inline komodo/release_transpiler.py:88]: the name `first_level_versions` points to something magical (what is the first level)? I see also that we don't assume which comes first. that's i think too liberal - i think we can pin down the sequence of coordinates, i think that is consistent with the whole "matrix" analogy. so perhaps rename to `rhel_versions`, or if you choose to rename `rhel_ver` to `rhel_coordinate`, then this could be renamed to `rhel_coordinate_values` - or if you wanna move to a more general thing, the arguments could be called `first_coord`, `second_coordinate`, and correspondingly `first_level_versions` could become `first_coordinate_values`.

**Review by valentin-krasontovitsch — CHANGES_REQUESTED** [review]
aaahhhh sooo many comments, i'm sorry. some of it, the typing stuff, is formalities. some of it is probably bike-shedding / arguing about naming. other things might be more substantial...

File: tests/data/test_release_matrix.yml
Comment (hnformentin) [reply tests/data/test_release_matrix.yml:28]: Interesting, if I only add a new line, I get the error:

> <stdin>:12: trailing whitespace.
>   warning: 1 line adds whitespace errors.

So I need to add a version for python 3.10 example.

Do you know how I should do to let it empty? Maybe only add `py310: ""` was required?

File: tests/test_release_transpiler.py
Comment (hnformentin) [reply tests/test_release_transpiler.py:29]: I think it is worth trying to write the test, at least it will add clarity to the functionality.
I will give it a try.

File: tests/test_release_transpiler.py
Comment (hnformentin) [reply tests/test_release_transpiler.py:54]: Good point!

File: tests/test_release_transpiler.py
Comment (hnformentin) [reply tests/test_release_transpiler.py:66]: nice, I need to think about it.

File: komodo/release_transpiler.py
Comment (valentin-krasontovitsch) [reply komodo/release_transpiler.py]: perfect, thank you!

File: komodo/release_transpiler.py
Comment (valentin-krasontovitsch) [reply komodo/release_transpiler.py:222]: > And to be honest I think it would be ok once we specify matrix-coordinates.

sure, sure - we can skip renaming the first one. don't wanna upset all the users of komodo out there : )

> But I did not play with it

i tried it out, sadly it doesn't display the default value - if the arg parse lib doesn't support it, you could of couse just pull out the default value into a variable...

File: tests/data/test_release_matrix.yml
Comment (valentin-krasontovitsch) [reply tests/data/test_release_matrix.yml]: hmm i think it's an editor thingy, this missing new line - i would hope that you don't need to specify an actual version, or an empty string or empty object (`{}` in yaml) for the last python version, while leaving the other python versions empty. who is it that complains?
sadly i never tried to understand the technical details around this "newline missing" complaint, so i don't know how to make it work exactly, but we can try to look at it together

File: komodo/release_transpiler.py
Comment (valentin-krasontovitsch) [reply komodo/release_transpiler.py]: > Do you agree with this one?

sure, sounds good!

ohhh i haven't worked with this feature yet from the coder point of view, so i'm not sure, probably you will have to do the rename in other places yourself.

File: komodo/release_transpiler.py
Comment (valentin-krasontovitsch) [reply komodo/release_transpiler.py:96]: after looking at the input we allow here, it becomes clear that more meaning cannot be added to the names, as we are way too permissive with the inputs we allow (I think). will create a separate issue to look at the input validation.

Comment (hnformentin) [reply komodo/release_transpiler.py:113]: We chatted that the better approach would be to include a longer comment, explaining that `pkg_versions` can take various formats.

**Review by hnformentin — COMMENTED** [review]
Some changes and corresponding replies so far!

File: tests/test_release_transpiler.py
Comment (valentin-krasontovitsch) [reply tests/test_release_transpiler.py:108]: looks good, thanks!

File: komodo/release_transpiler.py
Comment (valentin-krasontovitsch) [reply komodo/release_transpiler.py:241]: nice find! i'm still missing that it expects yaml formatted input in the help text. although i don't know how much it will help, and the default provides a working example, and makes it obvious how to extend it.

oh one point though that i forgot to mention before, you can remove the quotes around `rhel` and `python` in the default value, according to my testing at least

File: tests/test_release_transpiler.py
Comment (valentin-krasontovitsch) [reply tests/test_release_transpiler.py]: i don't see how you refactored the test, maybe it's lost in between different diffs, but i cannot see whether calling `build_matrix_file` without passsing the last argument is tested - have you covered that code path somewhere?

File: komodo/release_transpiler.py
Comment (hnformentin) [reply komodo/release_transpiler.py]: I changed the code to use the what you suggested, but actually did not understand how strings from errors exception work. I will add a comment on that on the new code.

File: tests/test_release_transpiler.py
Comment (hnformentin) [inline tests/test_release_transpiler.py:112]: If I stop the test at this point and `print(str(exception_info))`, I get:
`"'Rhel version rhel5 not found for package lib1.'"`
I don't understand why it has `'"` and how to remove `'`.
Do you have an idea for that?

Comment (valentin-krasontovitsch) [reply tests/test_release_transpiler.py:112]: oh ehm hmmm not sure - it's the quotation marks around the whole thing that you want to get rid of? how do you get the output / print ? do you use a debugger? is the test failing because of the quotation marks?

in general, i think you could match for a smaller part of the error message, to make the test less brittle when somebody should decide to change the error message - one could argue that we don't really care about the wording, but rather that some key information is present, like

```python
except KeyError as exception_info:
    for info in expected_info:
        assert info in str(exception_info)
```

where `expected_info` would for instance be `["py37", "rhel7","lib1"]` for the second test (and you could still keep the full text reason as the test id, for better understanding of what the test is testing / why it should fail)

Comment (hnformentin) [reply tests/test_release_transpiler.py]: Nice comment, I use the pattern all(), but have same effect.

Comment (valentin-krasontovitsch) [reply tests/test_release_transpiler.py]: great!

File: komodo/release_transpiler.py
Comment (valentin-krasontovitsch) [reply komodo/release_transpiler.py]: perfect, thank you!

**Review by valentin-krasontovitsch — APPROVED** [review]

**Comment by hnformentin** [comment]
This changes small changes after approval were made after discussion with the reviewer 🚀

**Pull request merged**
